File: aoc23/day-18/part2.py
from typing import List, Set, Dict, Tuple, Optional, Union, Iterable
import logging
import math
import structlog
from dataclasses import dataclass
from itertools import cycle
from enum import Enum, StrEnum
import sys
from shapely.geometry import Polygon
from shapely import Point

logger = structlog.get_logger()

class Direction(Enum):
    RIGHT = "R"
    LEFT = "L"
    UP = "U"
    DOWN = "D"

    @property
    def as_point(self) -> 'Point':
        return {
            Direction.RIGHT: Point(1, 0),
            Direction.LEFT: Point(-1, 0),
            Direction.UP: Point(0, -1),
            Direction.DOWN: Point(0, 1),
        }[self]


@dataclass()
class Instruction:
    direction: Direction
    color: Optional[int]
    distance: int

    def walk(self, start_point: Point) -> Point:
        logger.debug("walking", start_point=start_point, instruction=self)
        direction_point = self.direction.as_point

        return Point(direction_point.x * self.distance + start_point.x, direction_point.y * self.distance + start_point.y)


def part2(values_list) -> str:
    # structlog.contextvars.bind_contextvars(
    #     iteration=i,
    # )
    structlog.configure(
        wrapper_class=structlog.make_filtering_bound_logger(logging.DEBUG),
    )
    from structlog import get_logger
    logger = get_logger()
    logger.debug("part2")
    instructions = []
    d = []
    for line in values_list:
        color, _, distance_hex = line.split()
        distance_hex = distance_hex.removeprefix("(#").removesuffix(")")
        distance = int(distance_hex[:-1], 16)
        d.append(distance)
        direction = distance_hex[-1]
        match direction:
            case "0":
                direction = Direction.RIGHT
            case "1":
                direction = Direction.DOWN
            case "2":
                direction = Direction.LEFT
            case "3":
                direction = Direction.UP

        intruction = Instruction(direction=direction, distance=distance, color=color)
        logger.debug("instruction", instruction=intruction)
        instructions.append(intruction)

    start_point = Point(0, 0)
    current_point = start_point
    points = [start_point]
    for instruction in instructions:
        current_point = instruction.walk(current_point)
        points.append(current_point)
    else:
        logger.debug("done all instructions", current_point=current_point, points=points)

    geometry = Polygon(points)
    area = geometry.area
    perimeter = geometry.length
    result = int(perimeter/2 + area + 1)
    print(result)
    return f"{result}"

File: aoc23/day-18/src/part2.rs
// use geo::prelude::*;
// use geo::EuclideanDistance;
use geo::EuclideanLength;
use geo::{coord, Area, Coord, LineString, Polygon};
// use raqote::*;

enum Direction {
    North,
    South,
    East,
    West,
}

impl Direction {
    fn as_point(&self) -> Coord {
        match self {
            Direction::North => coord! {x: 0.0, y: -1.0},
            Direction::South => coord! {x: 0.0, y: 1.0},
            Direction::East => coord! {x: 1.0, y: 0.0},
            Direction::West => coord! {x: -1.0, y: 0.0},
        }
    }
}

struct Instruction {
    direction: Direction,
    distance: i32,
}

impl Instruction {
    fn walk(&self, start_point: Coord) -> Coord {
        let direction = self.direction.as_point();
        let distance = self.distance as f64;
        start_point + (direction * distance)
    }
}

pub fn process(input: &str) -> miette::Result<String> {
    let mut points = vec![];
    let mut current_point = coord! {x: 0.0, y: 0.0};
    points.push(current_point);
    for line in input.lines() {
        let instruction = parse_instruction(line);
        current_point = instruction.walk(current_point);
        points.push(current_point);
    }
    let line_string = LineString::new(points);
    // let min_x = line_string
    //     .points()
    //     .min_by(|a, b| a.x().partial_cmp(&b.x()).unwrap())
    //     .unwrap();
    // let min_y = line_string
    //     .points()
    //     .min_by(|a, b| a.y().partial_cmp(&b.y()).unwrap())
    //     .unwrap();
    // let max_x = line_string
    //     .points()
    //     .max_by(|a, b| a.x().partial_cmp(&b.x()).unwrap())
    //     .unwrap();
    // let max_y = line_string
    //     .points()
    //     .max_by(|a, b| a.y().partial_cmp(&b.y()).unwrap())
    //     .unwrap();

    // println!("min_x: {:?}", min_x);
    // println!("min_y: {:?}", min_y);
    // println!("max_x: {:?}", max_x);
    // println!("max_y: {:?}", max_y);

    let polygon = Polygon::new(line_string, vec![]);

    let perimeter = polygon.exterior().euclidean_length();
    let area = polygon.unsigned_area();
    let result = area + perimeter / 2.0 + 1.0;

    // let height = max_y.euclidean_distance(&min_y);
    // let width = max_x.euclidean_distance(&min_x);
    // println!("height: {:?}", height);
    // println!("width: {:?}", width);

    // let min_square = height.min(width);
    // let ratio = 2160.0 / min_square / 2.;
    // println!("ratio: {}", ratio);

    // polygon = polygon.scale(ratio);

    // println!("polygon: {:?}", polygon.unsigned_area());
    // let line_string = polygon.exterior().clone();
    // let min_x = line_string
    //     .points()
    //     .min_by(|a, b| a.x().partial_cmp(&b.x()).unwrap())
    //     .unwrap();
    // let min_y = line_string
    //     .points()
    //     .min_by(|a, b| a.y().partial_cmp(&b.y()).unwrap())
    //     .unwrap();
    // let max_x = line_string
    //     .points()
    //     .max_by(|a, b| a.x().partial_cmp(&b.x()).unwrap())
    //     .unwrap();
    // let max_y = line_string
    //     .points()
    //     .max_by(|a, b| a.y().partial_cmp(&b.y()).unwrap())
    //     .unwrap();
    // polygon = polygon.translate(min_x.x().abs() * ratio, min_y.y().abs() * ratio);

    // println!("min_x: {:?}", min_x);
    // println!("min_y: {:?}", min_y);
    // println!("max_x: {:?}", max_x);
    // println!("max_y: {:?}", max_y);

    // let min_x = line_string
    //     .points()
    //     .min_by(|a, b| a.x().partial_cmp(&b.x()).unwrap())
    //     .unwrap();
    // let min_y = line_string
    //     .points()
    //     .min_by(|a, b| a.y().partial_cmp(&b.y()).unwrap())
    //     .unwrap();
    // let max_x = line_string
    //     .points()
    //     .max_by(|a, b| a.x().partial_cmp(&b.x()).unwrap())
    //     .unwrap();
    // let max_y = line_string
    //     .points()
    //     .max_by(|a, b| a.y().partial_cmp(&b.y()).unwrap())
    //     .unwrap();
    // println!("min_x: {:?}", min_x);
    // println!("min_y: {:?}", min_y);
    // println!("max_x: {:?}", max_x);
    // println!("max_y: {:?}", max_y);

    // let mut dt = DrawTarget::new(3840, 2160);
    // let mut pb = PathBuilder::new();
    // for point in polygon.exterior().points() {
    //     pb.line_to(point.x() as f32, point.y() as f32);
    // }
    // let path = pb.finish();
    // let source = Source::Solid(SolidSource {
    //     r: 0x00,
    //     g: 0x00,
    //     b: 0x00,
    //     a: 0xff,
    // });
    // let _ = dt.fill(&path, &source, &DrawOptions::new());
    // let _ = dt.write_png("day18-part2.png");

    println!("Perimeter: {}", perimeter);
    println!("Area: {}", area);
    println!("Result: {}", result);
    Ok(format!("{}", result))
}

fn parse_instruction(line: &str) -> Instruction {
    // let mut (_, _, distance_hex) = line.split(' ');
    let mut split = line.splitn(3, ' ');
    let _old_direction_str = split.next().unwrap();
    let _old_distance_hex = split.next().unwrap();
    let mut hex_code = split.next().unwrap();
    hex_code = hex_code.trim_start_matches("(#");
    hex_code = hex_code.trim_end_matches(")");
    let mut split = hex_code.chars();
    let direction_str = split.next_back().unwrap();
    let distance_hex = split.as_str();
    let distance = i32::from_str_radix(distance_hex, 16).unwrap();

    let direction = match direction_str {
        '3' => Direction::North,
        '1' => Direction::South,
        '0' => Direction::East,
        '2' => Direction::West,
        _ => panic!("Invalid direction"),
    };
    Instruction {
        direction,
        distance,
    }
}
